### PR TITLE
Analysis tools made into ROOT dictionaries

### PR DESCRIPTION
**Pull request by siilieva**
The neutrino analysis tools are now added as ROOT dictionaries.
Two namespaces are introduced:  sndAnalysisCuts and sndAnalysisTools.
This modification was verbally discussed on several SoftPhys meetings and is further communicated with Cris as principle author of the tools. Other namespaces, or classes and functions inside the existing ones, can be added in a similar manner.

### CODE REVIEW

File: analysis/AnalysisToolsLinkDef.h
Comment (olantwin) [inline analysis/AnalysisToolsLinkDef.h:1]: If there is only one LinkDef in a directory, there is no need to give it a complicated name. We can just call them LinkDef.h (see [this related change in FairRoot](https://github.com/FairRootGroup/FairRoot/commit/222ac9a6a219a7d5e705f4b03c51c8aadfcc071a))

**Review by olantwin — COMMENTED** [review]
The last commit seems to undo most of the changes in the previous commits. I would just drop the first three commits to avoid confusion down the line and split the last commit into a fewer more easily digestable parts.

I'm not sure what the benefit of having two separate namespaces is, but maybe having an `Analysis` namespace with two nested namespaces `Cuts` and `Tools` might be better? And maybe we should generally use an `SND` namespace for everything?

I would suggest making one commit for all the renaming and then a second for the dictionary generation.

File: analysis/AnalysisToolsLinkDef.h
Comment (siilieva) [reply analysis/AnalysisToolsLinkDef.h:1]: I am not sure if we would like to add more linkdef files per directory in the case of analysis tools and cuts. We are just starting here. So I would keep the long names for now - in any case these only appear in cmake and the users won't be affected shall we change the linkdef filenames in the future.

**Comment by siilieva** [comment]
Oh, yes, the first commits are completely overwritten by the last one since I changed the approach - I will drop them. As for splitting the last commit - ok, still one of the new commits will change many files and be rather 'hard to digest'. 
>> maybe having an Analysis namespace with two nested namespaces Cuts and Tools might be better? And maybe we should generally use an SND namespace for everything?
The namespaces feature 'snd' : sndAnalysisCuts and sndAnalysisTools. I was envisioning AdvSND stuff will have _Adv_ prefix.
Why would be nested namespaces better?
I will close this request and make a new one.

File: analysis/AnalysisToolsLinkDef.h
Comment (olantwin) [reply analysis/AnalysisToolsLinkDef.h:1]: I think neither in FairRoot nor in sndsw has this ever been the case.

Why not start with a simple name used by convention in other projects as well, and in the off chance that we ever need to change it, it's a very simple change that the user won't notice ;)

**Comment by olantwin** [comment]
> Oh, yes, the first commits are completely overwritten by the last one since I changed the approach - I will drop them. As for splitting the last commit - ok, still one of the new commits will change many files and be rather 'hard to digest'.

Still, if it's only renames in many files, it's easy to skim. So I'd still prefer two commits. Maybe the rename first and then the linkdef, so that we don't need to rename everything in the linkdef.
 
> > > maybe having an Analysis namespace with two nested namespaces Cuts and Tools might be better? And maybe we should generally use an SND namespace for everything?
> 
> The namespaces feature 'snd' : sndAnalysisCuts and sndAnalysisTools. I was envisioning AdvSND stuff will have _Adv_ prefix. Why would be nested namespaces better? 

One of the nice things about namespaces (and nested namespaces) is that you can avoid all these cumbersome long names. We indeed should probably have an `adv` namespace as well.

With the nested namespaces, it would be easy to also load specific names (or everything from them). So you could use `using namespace SND` when you know you need many things from there.

Following [this](https://abseil.io/tips/130) and [this](https://google.github.io/styleguide/cppguide.html#Namespace_Names) I would probably prefer `snd::analysis_cuts` and `snd::analysis_tools` over `snd_analysis::{tools,cuts}` if you don't want to nest things too deep.

Since this are probably the first namespaces we introduce, we should also consider for a second, what naming convention we pick. snake_case seems more or less standard in C++ namespaces, but PascalCase also seems common. Not a huge fan of camelCase personally.

> I will close this request and make a new one.

Why?

**Comment by siilieva** [comment]
>>  So you could use using namespace SND when you know you need many things from there.

No, we avoid  _using namespace_ in sndsw like the general good practice, right? :D
Joke aside, I like the snd::analysis_cuts and snd::analysis_tools proposal. About the naming convention, online it is advised to stick with the _std_ classic _snake_case_ for namespaces.
No, need to close the request, true.

**Comment by olantwin** [comment]
> No, we avoid _using namespace_ in sndsw like the general good practice, right?

We shouldn't do it in headers, but no problem with doing it in the C++ files (especially for individual names).
Name resolution in C++ is generally [fun](https://en.cppreference.com/w/cpp/language/unqualified_lookup) though...

**Comment by siilieva** [comment]
The agreed changes are made - a bit more commit splitting was possible and so was done.
I suggest to use _snd:analysis_cuts:function_  throughout all code for consistency. At least I did it in for all executables in this request.

**Comment by olantwin** [comment]
Thanks!

I guess you weren't convinced about the LinkDef.h?

But generally looks good to me now (theoretically, I haven't tested the changes).

**Comment by siilieva** [comment]
I have tested using the neutrinoGolden selection - it works and produces the output of the current master sw.
I would like to have the branch merged so that people can adapt their code prior further development as we discussed during the SoftPhys meeting yesterday.

>>I guess you weren't convinced about the LinkDef.h?

Yes, I wasn't convinced. Once we are more advanced setting the analysis tools and see there are no multi-linkdef cases, we could have a separate commit renaming single linkdefs in all directories (/shipLHC, /analysis, et all).

**Comment by olantwin** [comment]
> > > I guess you weren't convinced about the LinkDef.h?
> 
> Yes, I wasn't convinced. Once we are more advanced setting the analysis tools and see there are no multi-linkdef cases, we could have a separate commit renaming single linkdefs in all directories (/shipLHC, /analysis, et all).

Deal.

**Review by olantwin — APPROVED** [review]

**Pull request merged**